Enable platform-specific binary package filtering input

taylormadore opened the pull request:
This PR introduces flexible binary package filtering input parameters that allow users to specify which platform-specific packages to prefetch based on architecture, OS, Python version, and other characteristics.

This is **only** the input model implementation and does not change the current behavior. New behavior for the individual package manager backends will be added in subsequent PRs.

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

taylormadore commented:
@coderabbitai review

slimreaper35 commented on hermeto/core/models/input.py:
```diff
@@ -120,11 +153,113 @@ def _validate_ssl_options(self) -> Self:
+        )
+
+    value = value.strip()
+    if value == BINARY_FILTER_ALL:
```
I see uppercase `:ALL:` in the unit tests, but here is a check only for the lowercase value.

taylormadore replied:
Currently the magic `:all:` keyword is case-sensitive, so I think that is the expected behavior. The test case is showing that `:ALL:` is accepted from an input standpoint but does not trigger the same behavior as `:all:`.

slimreaper35 commented:
I don't think I have any major comments 👍 Hopefully, I didn't miss something.

eskultety commented on hermeto/core/models/input.py:
```diff
@@ -120,6 +122,60 @@ def _validate_ssl_options(self) -> Self:
+    if value == BINARY_FILTER_ALL:
+        return BinaryFilter()
+
+    filters = {f.strip() for f in value.split(",") if f.strip()}
```
nitpick: can we use walrus here? `{f for f in value.split(",") if (f := f.strip())}` (or with any other variable other than `f` :) ).

taylormadore replied:
Done

eskultety commented:
> This is **only** the input model implementation and does not change the current behavior. New behavior for the individual package manager backends will be added in subsequent PRs.

Ah, so this means that

> or python implementation, the implementation-agnostic py and ...

will be done in a follow up, correct?

a-ovchinnikov commented on hermeto/core/models/input.py:
```diff
@@ -120,6 +122,60 @@ def _validate_ssl_options(self) -> Self:
+        )
+
+
+def _parse_binary_filter(value: Any) -> BinaryFilter:
```
This feels like a classmethod for BinaryFilter, is there any reason to keep this separate from the class?

a-ovchinnikov commented on hermeto/core/models/input.py:
```diff
@@ -120,6 +122,60 @@ def _validate_ssl_options(self) -> Self:
+    def __repr__(self) -> str:
+        return (
+            f"<{type(self).__name__} "
+            f"filters={':all:' if self.is_all else ','.join(sorted(self.filters))}>"
```
The [recommended way](https://docs.python.org/3/reference/datamodel.html#object.__repr__) of defining a \_\_repr\_\_() is to have it return something that
```
If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value
```
I.e. here you would want to have something like `BinaryFilter(filters=set())`. Maybe this is a sign to move `:all:` processing into the initializer?

a-ovchinnikov commented on tests/unit/models/test_input.py:
```diff
@@ -393,3 +397,160 @@ def test_empty_packages(self) -> None:
+            pytest.param("   ", "No valid filters found", id="only_whitespace"),
+            pytest.param(",,,", "No valid filters found", id="only_commas"),
+            pytest.param(" , , ", "No valid filters found", id="comma_whitespace_only"),
+        ],
+    )
+    def test_parse_string_invalid_cases(self, input_value: str, error_match: str) -> None:
+        with pytest.raises(ValueError, match=error_match):
+            _parse_binary_filter(input_value)
+
+    @pytest.mark.parametrize(
+        "invalid_type",
+        [
+            pytest.param(123, id="integer"),
+            pytest.param(["x86_64"], id="list"),
+            pytest.param({"arch": "x86_64"}, id="dict"),
+            pytest.param(True, id="boolean"),
+        ],
+    )
+    def test_parse_invalid_types(self, invalid_type: Any) -> None:
+        with pytest.raises(ValueError) as exc_info:
+            _parse_binary_filter(invalid_type)
+        error_msg = str(exc_info.value)
+        assert f"Got type: {type(invalid_type).__name__}" in error_msg
+        assert "must be a string" in error_msg
+        assert "e.g., 'x86_64,aarch64' or ':all:'" in error_msg
+
+    @pytest.mark.parametrize(
+        "input_value,expected_filters",
+        [
+            pytest.param("x86_64,café,日本", {"x86_64", "café", "日本"}, id="unicode_values"),
+            pytest.param("x86-64,arm-linux", {"x86-64", "arm-linux"}, id="hyphenated_values"),
+            pytest.param(
+                "linux_gnu,darwin_x64", {"linux_gnu", "darwin_x64"}, id="underscore_values"
+            ),
+            pytest.param("3.9,3.10,3.11", {"3.9", "3.10", "3.11"}, id="dotted_values"),
+            pytest.param("x86_64-linux-gnu", {"x86_64-linux-gnu"}, id="mixed_special_chars"),
+            pytest.param("123,456,789", {"123", "456", "789"}, id="numeric_string_values"),
+            pytest.param("x86_64,123,abc", {"x86_64", "123", "abc"}, id="mixed_numeric_alpha"),
+        ],
+    )
```
It looks like we are testing Python stdlib here, do we need to do that?

taylormadore replied:
Tests substantially reworked and reduced in V2

a-ovchinnikov commented on hermeto/core/models/input.py:
```diff
@@ -23,6 +23,37 @@
+    elif should_migrate_allow_binary:
+        log.warning(
+            "The 'allow_binary' field is deprecated and will be removed in the next major version. "
+            "Please use 'binary': {} instead of 'allow_binary': true."
```
Should it be something like `'binary': {'arch...}` or do you mean that an empty map is fine too?

a-ovchinnikov commented on tests/unit/models/test_input.py:
```diff
@@ -554,3 +655,111 @@ def test_field_accepts_none(self) -> None:
+
+
+class TestPipBinaryFilters:
+    def test_default_values(self) -> None:
```
Here we test that pydantic sets default values to the values that we asked it to use. Which means that the test is for pydantic. Does it provide value to Hermeto?

a-ovchinnikov reviewed (approved):
LGTM with several nitpicks and optional suggestions.

taylormadore commented:
Changes in V2:
- Kept the input validation focused on the input string format. I felt like I was trying to do too much parsing there in V1 and abusing the validator
- Cut down and simplified the unit tests
- Added the RPM arch filtering as an example (to be dropped from this before merge and opened separately)

slimreaper35 commented:
Sigh, I didn't expand the commit message 😞

taylormadore commented on hermeto/core/models/input.py:
```diff
@@ -23,6 +23,37 @@
+    instance: Union["PipPackageInput", "BundlerPackageInput"],
+    binary_filter_class: Union[type["PipBinaryFilters"], type["BundlerBinaryFilters"]],
+) -> None:
+    """Handle backward compatibility for allow_binary field."""
```
I agree that should be mentioned for this one. Will fix :+1:

taylormadore commented on hermeto/core/models/input.py:
```diff
@@ -140,6 +142,27 @@ def _validate_ssl_options(self) -> Self:
+def _validate_binary_filter_format(value: Any) -> str:
+    """Validate binary filter format as either a single value or comma-separated string.
+
+    May mutate the input value slightly to return non-empty, unique, trimmed values
```
What I wanted to convey is that this validator isn't strictly validating the input--it _**may**_ also make some small changes to normalize the input string

taylormadore commented on tests/unit/models/test_input.py:
```diff
@@ -393,3 +397,160 @@ def test_empty_packages(self) -> None:
+        assert filter.is_all is expected_is_all
+
+
+class TestParseBinaryFilter:
```
Tests substantially reworked in V2
